### PR TITLE
doc: clarify release guide

### PR DESCRIPTION
**Pull request by allenlsy**
**Description**:

The changes here clarify the process of creating patch branch, and
merging it to release branch.

Signed-off-by: Allen Leigh <allenlsy@gmail.com>

**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| New Functionality          | [ ] |
| Documentation              | [X] |
| Install                    | [ ] |
| CLI Tool                   | [ ] |
| Certificate Management     | [ ] |
| Control Plane              | [ ] |
| Ingress                    | [ ] |
| Egress                     | [ ] |
| Networking                 | [ ] |
| Observability              | [ ] |
| SMI Policy                 | [ ] |
| Sidecar Injection          | [ ] |
| Security                   | [ ] |
| Upgrade                    | [ ] |
| Tests                      | [ ] |
| CI System                  | [ ] |
| Demo                       | [ ] |
| Performance                | [ ] |
| Other                      | [ ] |

Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project?
    -   Did you notify the maintainers and provide attribution?

No

1. Is this a breaking change?

No

### CODE REVIEW

File: docs/release_guide.md [hunk 1]
```diff
@@ -43,13 +42,15 @@ Push the release branch to the main repo (not fork), identified here by the `ups
 $ git push upstream release-<version> # ex: git push upstream release-v0.4
 ```
 
+Note that, DO NOT create a new branch for RCs in the upstream repo.
```
Comment (snehachhabria) [inline docs/release_guide.md:45]:
```suggestion
_Note:_ Do not create a new branch for RCs in the upstream repo.
```

File: docs/release_guide.md [hunk 2]
```diff
@@ -27,7 +27,6 @@ Once an RC has been found to be stable, cut a release tagged `vX.Y.Z` using the
   - [Update docs site](#update-docs-site)
   - [Announce the new release](#announce-the-new-release)
   - [Make version changes on main branch](#make-version-changes-on-main-branch)
-  - [Make version changes on docs.openservicemesh.io](#make-version-changes-on-docsopenservicemeshio)
```
Comment (snehachhabria) [inline docs/release_guide.md:30]: Changes need to be made in the docs repo (osm-docs), please could you add a step for that and the relevant files if possible

File: docs/release_guide.md [hunk 3]
```diff
@@ -121,3 +123,28 @@ Make an announcement on the [mailing list](https://groups.google.com/g/openservi
 Skip this step if the release is a release candidate (RC).
 
 Open a pull request against the `main` branch making the same version updates as [above](#update-release-branch-with-patches-and-versioning-changes) so the latest release assets are referenced there.
+
+## Make version changes on docs.openservicemesh.io
+
+Associated version numbers need to be updated in the [docs.openservicemesh.io](https://github.com/openservicemesh/osm-docs/) repo.
+
+The files needs to be updated are:
```
Comment (jaellio) [inline docs/release_guide.md:131]: nit:

```suggestion
The files that need to be updated are:
```

File: docs/release_guide.md [hunk 4]
```diff
@@ -121,3 +123,28 @@ Make an announcement on the [mailing list](https://groups.google.com/g/openservi
...
 
 Open a pull request against the `main` branch making the same version updates as [above](#update-release-branch-with-patches-and-versioning-changes) so the latest release assets are referenced there.
+
+## Make version changes on docs.openservicemesh.io
+
+Associated version numbers need to be updated in the [docs.openservicemesh.io](https://github.com/openservicemesh/osm-docs/) repo.
+
+The files that need to be updated are:
+
+* [Readme](https://github.com/openservicemesh/osm-docs/blob/main/README.md)
```
Comment (nojnhuh) [inline docs/release_guide.md:133]: What exactly needs to be updated here? I only see references to release branches as an example, but those don't need to stay up to date.

File: docs/release_guide.md [hunk 5]
```diff
@@ -121,3 +123,28 @@ Make an announcement on the [mailing list](https://groups.google.com/g/openservi
 Skip this step if the release is a release candidate (RC).
 
 Open a pull request against the `main` branch making the same version updates as [above](#update-release-branch-with-patches-and-versioning-changes) so the latest release assets are referenced there.
+
+## Make version changes on docs.openservicemesh.io
```
Comment (nojnhuh) [inline docs/release_guide.md:127]: We should add a note NOT to do this step for pre-releases (which includes RCs).

File: docs/release_guide.md [hunk 6]
```diff
@@ -121,3 +123,28 @@ Make an announcement on the [mailing list](https://groups.google.com/g/openservi
...
+
+Associated version numbers need to be updated in the [docs.openservicemesh.io](https://github.com/openservicemesh/osm-docs/) repo.
+
+The files that need to be updated are:
+
+* [Readme](https://github.com/openservicemesh/osm-docs/blob/main/README.md)
+* [Manual Demo](https://github.com/openservicemesh/osm-docs/blob/main/content/docs/getting_started/manual_demo.md)
+* [Health Probes](https://github.com/openservicemesh/osm-docs/blob/main/content/docs/application_health_probes.md)
```
Comment (nojnhuh) [inline docs/release_guide.md:135]: The only references to versions I see here are in logs, so I don't think we need to update those every time we cut a release.

File: docs/release_guide.md [hunk 7]
```diff
@@ -121,3 +123,28 @@ Make an announcement on the [mailing list](https://groups.google.com/g/openservi
...
+* [Control Plane Health Probes](https://github.com/openservicemesh/osm-docs/blob/main/content/docs/control_plane_health_probes.md)
+* [OSM MeshConfig](https://github.com/openservicemesh/osm-docs/blob/main/content/docs/osm_mesh_config.md)
+* [Upgrade Guide](https://github.com/openservicemesh/osm-docs/blob/main/content/docs/upgrade_guide.md)
```
Comment (nojnhuh) [inline docs/release_guide.md:138]: I only see versions referenced here as an example.

File: docs/release_guide.md [hunk 8]
```diff
@@ -121,3 +123,28 @@ Make an announcement on the [mailing list](https://groups.google.com/g/openservi
 Skip this step if the release is a release candidate (RC).
 
 Open a pull request against the `main` branch making the same version updates as [above](#update-release-branch-with-patches-and-versioning-changes) so the latest release assets are referenced there.
+
+## Make version changes on docs.openservicemesh.io
+
+Associated version numbers need to be updated in the [docs.openservicemesh.io](https://github.com/openservicemesh/osm-docs/) repo.
+
+The files that need to be updated are:
```
Comment (nojnhuh) [inline docs/release_guide.md:131]: I only looked at the first few files here, but several seemed like they don't need to be included. Please take another look at these to trim it down as much as possible.

File: docs/release_guide.md [hunk 9]
```diff
@@ -121,3 +123,28 @@ Make an announcement on the [mailing list](https://groups.google.com/g/openservi
...
+
+* [Readme](https://github.com/openservicemesh/osm-docs/blob/main/README.md)
+* [Manual Demo](https://github.com/openservicemesh/osm-docs/blob/main/content/docs/getting_started/manual_demo.md)
+* [Health Probes](https://github.com/openservicemesh/osm-docs/blob/main/content/docs/application_health_probes.md)
+* [Control Plane Health Probes](https://github.com/openservicemesh/osm-docs/blob/main/content/docs/control_plane_health_probes.md)
+* [OSM MeshConfig](https://github.com/openservicemesh/osm-docs/blob/main/content/docs/osm_mesh_config.md)
+* [Upgrade Guide](https://github.com/openservicemesh/osm-docs/blob/main/content/docs/upgrade_guide.md)
+* [Installation](https://github.com/openservicemesh/osm-docs/blob/main/content/docs/install/_index.md)
```
Comment (nojnhuh) [inline docs/release_guide.md:139]: Where is there a version here at all?

**Review by nojnhuh — CHANGES_REQUESTED** [review]

File: docs/release_guide.md (hunk 9 above)
Comment (allenlsy) [reply docs/release_guide.md:137]: It's the link of OSM Chart:

https://github.com/openservicemesh/osm-docs/blob/main/content/docs/install/_index.md#using-the-helm-cli

File: docs/release_guide.md [hunk 10]
```diff
@@ -121,3 +123,18 @@ Make an announcement on the [mailing list](https://groups.google.com/g/openservi
...
+
+## Make version changes on docs.openservicemesh.io
+
+**Note**: do not perform this step for pre-releases.
+
+Associated version numbers need to be updated in the [docs.openservicemesh.io](https://github.com/openservicemesh/osm-docs/) repo.
+
+The files that need to be updated are:
+
+* [Control Plane Health Probes](https://github.com/openservicemesh/osm-docs/blob/main/content/docs/control_plane_health_probes.md)
```
Comment (nojnhuh) [inline docs/release_guide.md:135]: References to versions here are just sample output. I'd omit this file.

File: docs/release_guide.md (hunk 9 above)
Comment (nojnhuh) [reply docs/release_guide.md:137]: Ah thanks, I see it now.

File: docs/release_guide.md (hunk 8 above)
Comment (allenlsy) [reply docs/release_guide.md:133]: I further trimmed the list down. The criteria is, we only update the docs that version number affects sample shell commands, or links to the latest code.

Comment (nojnhuh) [reply docs/release_guide.md:133]: Thanks for doing that, I think some should be added back though. I'm going through https://github.com/openservicemesh/osm-docs/pull/109 to see if there's anything missing there. That PR changed a few spots that we don't need to update every time, but I'll leave comments on each of those spots. Then let's make this list based off of the changes there. I'll let you know when I'm finished reviewing that PR.

Sorry this is turning into a bigger deal than it probably should, but making this list complete and concise should save us headaches down the road.

Comment (allenlsy) [reply docs/release_guide.md:133]: @nojnhuh I think osm-docs #109 is finished. I will merge this PR for v0.9.x release

Comment (ksubrmnn) [reply docs/release_guide.md:133]: Is it worth including a list here? Or should we guide devs to just use the search function?

Comment (allenlsy) [reply docs/release_guide.md:133]: I actually think so. We do the same, and the list is not always accurate.

Comment (allenlsy) [reply docs/release_guide.md]: @nojnhuh I removed the concrete files list here.

File: docs/release_guide.md [hunk 11]
```diff
@@ -38,22 +38,23 @@ Identify the base commit in the `main` branch for the release and cut a release
...
 ```console
 $ git push upstream release-<version> # ex: git push upstream release-v0.4
 ```
 
+_Note_: Do not create a new branch for RCs in the upstream repo.
+
 ## Update release branch with patches and versioning changes
 
-Create a new branch off of the release branch to maintain updates specific to the new version.
+Create a new branch off of the release branch to maintain updates specific to the new version. Let's call it the patch branch. The patch branch should not be created in the upstream repo.
```
Comment (nojnhuh) [inline docs/release_guide.md:50]: > The patch branch should not be created in the upstream repo.

If we delete the branch after its PR is merged, I don't see why it makes a difference where the branch lives. I understand every branch is currently captured by the upstream repo's branch protection rules which means an admin has to delete the branch and that's cumbersome.

I think a better solution to that particular problem is to fix in the branch protection rules vs. qualifying every current and future instruction about pushing short-lived branches with this same statement. Keeping this seems fine to me until then.

File: docs/release_guide.md [hunk 12]
```diff
@@ -93,6 +94,10 @@ It will build the CLI binaries, publish a new GitHub release,
 upload the packaged binaries and checksums as release assets, build and push Docker images for OSM and the demo to the
 [`openservicemesh` organization](https://hub.docker.com/u/openservicemesh) on Docker Hub, and publish the Helm chart to the repo hosted at https://openservicemesh.github.io/osm.
 
+To summarize, here is a diagram demonstrate the git branching during the release process.
+
+![](../img/osm-git-release.jpg)
```
Comment (nojnhuh) [inline docs/release_guide.md:99]: Personally, I find this diagram a bit overwhelming and hard to follow. This seems like a symptom that something in the text could be made more clear, in which case I'd rather clarify the text than supplement it with a diagram in the spirit of trying to keep this document as lean and easy to follow as we reasonably can.

I'm certainly open to keeping it if others find something like this helpful, but it does need at least one change:

> 3. merge patch to release branch

At this point, there have been no changes added to the patch branch. The changes from steps 4 and 5 can be made on the patch branch then a single PR can be made from the patch branch to the release branch.

File: docs/release_guide.md [hunk 13]
```diff
@@ -121,3 +126,11 @@ Make an announcement on the [mailing list](https://groups.google.com/g/openservi
...
 
 Open a pull request against the `main` branch making the same version updates as [above](#update-release-branch-with-patches-and-versioning-changes) so the latest release assets are referenced there.
+
+## Make version changes on docs.openservicemesh.io
+
+**Note**: do not perform this step for pre-releases.
+
+Associated version numbers need to be updated in the [docs.openservicemesh.io](https://github.com/openservicemesh/osm-docs/) repo.
+
+The suggested way is to search for hard coded current version number in the osm-docs repo, then update the ones that are will brake users' experience with if they follow the documentation, such as demonstration commands, reference links and anything that is strongly related to the next release. We don't need to update the version numbers that just serve the purpose of examples.
```
Comment (nojnhuh) [inline docs/release_guide.md:136]: - Some references to code on release branches should be changed, so add a mention to also search for the release branch name
- grammar nits
- add a link to https://github.com/openservicemesh/osm-docs/pull/109 as an example.

```suggestion
The suggested way is to search for hard coded instances of the current version number and release branch name in the osm-docs repo, then update the ones that may break users' experiences if they follow the documentation, such as demonstration commands, reference links and anything that is strongly related to the next release. We don't need to update the version numbers that just serve the purpose of examples.

See https://github.com/openservicemesh/osm-docs/pull/109 for an example of the update from v0.8.4 to v0.9.0.
```

**Review by nojnhuh — CHANGES_REQUESTED** [review]

File: docs/release_guide.md [hunk 14]
```diff
@@ -38,22 +38,23 @@ Identify the base commit in the `main` branch for the release and cut a release
 $ git checkout -b release-<version> <commit-id> # ex: git checkout -b release-v0.4 0d05587
 ```
 
-Push the release branch to the main repo (not fork), identified here by the `upstream` remote.
+Push the release branch to the upstream repo (not forked), identified here by the `upstream` remote.
 ```console
 $ git push upstream release-<version> # ex: git push upstream release-v0.4
 ```
 
+_Note_: Do not create a new branch for RCs in the upstream repo.
```
[the remark this answers is not in the document]
Comment (allenlsy) [reply docs/release_guide.md:46]: I get you point. The emphasis of this line is "in the upstream repo".

File: docs/release_guide.md (hunk 12 above)
Comment (allenlsy) [reply docs/release_guide.md]: Discussed offline. Diagram removed for simplicity.

File: docs/release_guide.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (allenlsy) [reply docs/release_guide.md]: Line removed.

File: docs/release_guide.md (hunk 12 above)
Comment (nojnhuh) [reply docs/release_guide.md]: Certainly not necessary to handle in this PR, but another thought I had was to add a diagram that shows only the end result of the commits/branches/tags, since that's what's most important. Exactly how that result is achieved (PRs from certain branches, cherry-picking, merging) isn't all that important and may change, but it's good to document at least one method documented with concrete steps. I don't think the equivalent of that is described in words anywhere and would be useful context to have while performing the steps.

Again, I don't mean to dictate for everybody what is or isn't helpful, but I think it's important to avoid ambiguity in this doc as much as possible.

@allenlsy could you at least add the diagram back in a comment on this thread so we can get some more feedback?

**Review by nojnhuh — APPROVED** [review]

**Comment by allenlsy** [comment]
Link to issue https://github.com/openservicemesh/osm/pull/3533.
